Adiciona post sobre CSS Grid (#10)
OpenDevUFCG/OpenDevUFCGBlog: texts/css-grid.md

---
title: Criando malhas simples com CSS Grid
published: false
description: Introdução a CSS grid
tags: css, grid, ptbr, tips

> series: grid-tips

Reviewer: ```suggestion
```
O nome de uma série é por extenso mesmo (por exemplo, CSS Grid ou Dicas de CSS Grid)... Mas minha sugestão é só editar para colocar que é parte de uma série quando for fazer outras postagens futuramente. ;)

---

# Grid para ser feliz

> Você conhece o CSS Grid? Se não, você está perdendo uma otima ferramenta do CSS que facilita tanto o desenvolvimento como tambem a manutenção. Nesse post, irei tratar um pouco de como ele funciona e como ele pode ser poderoso.

Reviewer: ```suggestion
Você conhece o CSS Grid? Se não, você está perdendo uma ótima ferramenta do CSS que facilita tanto o desenvolvimento como também a manutenção de código. Nesse post, irei tratar um pouco de seu funcionamento e de como ele pode ser poderoso.
```

# Um pouco sobre caixas

> Pra começar você tem que saber que todo elemento HTML está envolvido por uma caixa que delimita o elemento HTML e o seu conteudo ([para saber mais acesse esse link](https://tableless.github.io/iniciantes/manual/css/box-model.html) e [esse link](https://developer.mozilla.org/pt-BR/docs/Web/CSS/box_model)). Na maioria das vezes, vão existir elementos que tem dentro do seu conteudo outros elementos HTML, ou seja, caixas dentro de caixas, e o que o Grid faz é simplesmente organizar essas caixas de acordo com sua vontade.

Reviewer: ```suggestion
Pra começar, você tem que abstrair que todo elemento HTML está envolvido por uma caixa que tanto o delimita como também o seu conteúdo (para saber mais, acesse [esse link](https://tableless.github.io/iniciantes/manual/css/box-model.html) e [esse link](https://developer.mozilla.org/pt-BR/docs/Web/CSS/box_model)). Na maioria das vezes, vão existir elementos que têm dentro do seu conteúdo outros elementos HTML, ou seja, caixas dentro de caixas, e o que o Grid faz é simplesmente organizar essas caixas de acordo com sua vontade.
```

Aqui eu mudei "entender" para "abstrair" já que é mais uma analogia para entender. Certo?

Reviewer: Eu também juntaria a linha que está a seguir "Como ele faz isso?..." Aqui, já que é parte da mesma noção que você está trabalhando no parágrafo.

Como ele faz isso? Ele define uma malha quadriculada e atribui espaços dessa malha a cada elemento HTML.

![](https://thepracticaldev.s3.amazonaws.com/i/58gv9hkeys3de94e0amo.png)

> A malha é definida por linhas e colunas que são medidas em `fr`. A medida `fr` foi criada exclusivamente para o CSS Grid e pode ser definida como `Uma fração do espaço disponível no container do grid`(Fonte: [MDN](https://developer.mozilla.org/pt-BR/docs/Web/CSS/CSS_Grid_Layout/Basic_Concepts_of_Grid_Layout)). No caso da imagem acima, as colunas tem a medida de `5fr` e as linhas de `3fr`. E sabe o melhor do `fr`? Ele se adequa ao tamanho da sua tela. 

Reviewer: ```suggestion
A malha é definida por linhas e colunas que são medidas em `fr`. A medida `fr` foi criada exclusivamente para o CSS Grid e pode ser definida como "uma fração do espaço disponível no container do grid" (fonte: [MDN](https://developer.mozilla.org/pt-BR/docs/Web/CSS/CSS_Grid_Layout/Basic_Concepts_of_Grid_Layout)). No caso da imagem acima, as colunas têm a medida de `5fr` e as linhas de `3fr`. E sabe o melhor do `fr`? Ele se adequa ao tamanho da sua tela. 
```

> Além disso cada caixa pintada da imagem é uma área do grid onde ficará um elemento HTML.

Reviewer: ```suggestion
Além disso, cada caixa pintada da imagem é uma área do grid onde ficará um elemento HTML.
```
E mesma coisa, eu juntaria com o parágrafo acima.

# Colocando em prática

## Preparando o terreno

> O primeiro passo para utilizar o grid é definir quais são os elementos que estamos utilizando e o que cada um contem. Isso é definido no HTML, e para facilitar o entendimento, usarei tags HTML personalizadas. Segue o exemplo:

Reviewer: ```suggestion
O primeiro passo para utilizar o grid é definir quais são os elementos que estamos utilizando e o que cada um contém. Isso é definido no HTML e, para facilitar o entendimento, usarei tags HTML personalizadas. Segue o exemplo:
```

```html
<caixa-grande>
  <caixa-pequena1/>
  <caixa-pequena2/>
  <caixa-pequena3/>
  <caixa-pequena4/>
  <caixa-pequena5/>
  <caixa-pequena6/>
</caixa-grande>
```

> Nesse caso, temos uma caixa grande que tem seis caixas pequenas dentro dela. Para começar a utilização do grid você precisa apenas definir a propriedade `display: grid` no elemento que você desejar: 

Reviewer: ```suggestion
Nesse caso, temos uma caixa grande que tem seis caixas pequenas dentro dela. Para começar a utilização do grid, você precisa apenas definir a propriedade `display: grid` no elemento que desejar: 
```

```css
caixa-grande {
  display: grid;
}

```

> Facil, né? Mas calma, ainda não temos nada pronto. O que foi definido nesse CSS foi apenas o tipo de organização (`display`) que a caixa utilizará, que é o grid. Agora precisamos organiza-la de fato.

Reviewer: ```suggestion
Facil, né? Mas calma, ainda não temos nada pronto. O que foi definido nesse CSS foi apenas o tipo de organização (`display`) que a caixa utilizará, que é o grid. Agora precisamos organizá-la de fato.
```

# Entendendo como organizar

> Para isso você precisa definir o grid - ou seja, a malha - no elemento HTML que você escolheu. Nessa malha você vai definir tres coisas: as colunas, as linhas e as áreas para colocar o conteudo da caixa que você escolher.

Reviewer: ```suggestion
Para isso, você precisa definir o grid — ou seja, a malha — no elemento HTML de sua escolha. Nessa malha você vai definir três coisas: as colunas, as linhas e as áreas para colocar o conteúdo da caixa que você escolher.
```

> Essa organização em código se dá pela definição do `grid-template`. Essa propriedade pode ser dividida em outras tres para facilitar a sua declaração. São elas:

Reviewer: ```suggestion
Essa organização em código se dá pela definição do `grid-template`. Essa propriedade pode ser dividida em outras três para facilitar a sua declaração. São elas:
```

> 1. `grid-template-columns`: Define as colunas

Reviewer: Isso, é bom esclarecer.

> 1. `grid-template-columns`: Define as colunas

Reviewer: ```suggestion
1. `grid-template-columns`: Define as colunas;
```

> 1. `grid-template-rows`: Define as linhas

Reviewer: ```suggestion
2. `grid-template-rows`: Define as linhas;
```

> 1. `grid-template-areas`: Define as áreas

Reviewer: ```suggestion
3. `grid-template-areas`: Define as áreas.
```

> Você não precisa definir todas essas propriedades ao mesmo tempo, o grid é inteligente o bastante para deduzir algumas coisas.

Reviewer: ```suggestion
Você não precisa definir todas essas propriedades ao mesmo tempo; o grid é inteligente o bastante para deduzir algumas coisas.
```

## Codificando

> Vamos para um exemplo concreto. Imagine que você queira todas as caixas, uma do lado da outra na horizontal. Muito simples:

Reviewer: ```suggestion
Vamos para um exemplo concreto. Imagine que você queira todas as caixas uma do lado da outra na horizontal. Muito simples:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
}
```

> Ou para facilitar:

Reviewer: ```suggestion
Ou, para facilitar:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: repeat(6, 1fr);
}
```

> O repeat é utilizado para repetir varias vezes a quantidade de medida passada. Nesse caso, repetirmos seis vezes a medida de `1fr`. O repeat é uma função muito poderosa e será abordada em outro post.

Reviewer: ```suggestion
O `repeat` é utilizado para repetir várias vezes a quantidade de medida passada. Nesse caso, repetimos seis vezes a medida de `1fr`. O repeat é uma função muito poderosa e será abordada em outro post.
```

E se quisermos um abaixo do outro? Simples também:

```css
caixa-grande {
  display: grid;
  grid-template-rows: repeat(6, 1fr);
}
```

Agora se quisermos uma malha 3x2? Continua muito simples:

```css
caixa-grande {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: repeat(2, 1fr);
}
```

> Em todos esses casos tanto as colunas como as linhas estão com tamanhos iguais, mas podemos definir com tamanhos diferentes também, por exemplo:

Reviewer: ```suggestion
Em todos esses casos, tanto as colunas como as linhas estão com tamanhos iguais, mas podemos definir com tamanhos diferentes também. Por exemplo:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: 2fr repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);
}
```

> Agora a primeira coluna tem o dobro do tamanho das outras colunas, simples ne?

Reviewer: ```suggestion
Agora a primeira coluna tem o dobro do tamanho das outras colunas. Simples, né?
```

> ## Usando areas no grid

Reviewer: ```suggestion
## Usando áreas no grid
```

> Em nenhum desses exemplos utilizamos o `grid-template-areas`, no entanto todos funcionaram bem. Funcionam justamente por causa da dedução do CSS Grid, pois ele atribui sequencialmente as propriedades da malha a cada filho.

Reviewer: ```suggestion
Em nenhum desses exemplos utilizamos o `grid-template-areas`, no entanto, todos funcionaram bem. Funcionam justamente por causa da dedução do CSS Grid que mencionei anteriormente, pois ele atribui sequencialmente as propriedades da malha a cada filho.
```

> No entanto caso queiramos alterar a ordem das caixas, precisamos alterar o HTML. Num exemplo simples como esse basta trocar a ordem das linhas das caixas e cada conteudo será alterado, no entanto em projetos complexos se torna inviavel alterar o HTML. Para a nossa sorte existe o `grid-template-areas`.

Reviewer: ```suggestion
No entanto, caso queiramos alterar a ordem das caixas, precisamos alterar o HTML. Num exemplo simples como esse, basta trocar a ordem das linhas das caixas e cada conteúdo será alterado. No entanto, em projetos complexos, se torna inviável alterar o HTML. Para a nossa sorte existe o `grid-template-areas`.
```

> Nele você define, literalmente, áreas que a malha terá. Como exemplo:

Reviewer: Juntar com o parágrafo acima.

```css
caixa-grande {
  display: grid;
  grid-template-columns: 2fr repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-template-areas: 
    "c-pequena1 c-pequena2 c-pequena3"
    "c-pequena4 c-pequena5 c-pequena6";
}
```

> O container pai está pronto, mas precisamos ainda definir dentro de cada filho qual area ele é atribuido, desse modo:

Reviewer: ```suggestion
O container pai está pronto, mas precisamos ainda definir dentro dos filhos a qual área cada um está atribuído, o que é feito desse modo:
```

```css

> caixa-pequena1 {

Reviewer 2: Achei legal mostrar isso. Teria como tu colocar uma cor pra cada área e mostrar como ficaria em tela? O devto também tem a funcionalidade de renderizar um codepen no meio do MD. Acho que ficaria muito legal a pessoa ver o código, poder modificar e ainda ver na prática

  grid-area: "c-pequena1";
}

caixa-pequena2 {
  grid-area: "c-pequena2";
}

caixa-pequena3 {
  grid-area: "c-pequena3";
}

caixa-pequena4 {
  grid-area: "c-pequena4";
}

caixa-pequena5 {
  grid-area: "c-pequena5";
}

caixa-pequena6 {
  grid-area: "c-pequena6";
}
```

> Tudo pronto! Essa definição ficou igual ao exemplo anterior onde não tinhamos o `grid-template-areas`, mas imagine que queiramos trocar a posição da `caixa-pequena2` pela `caixa-pequena4`. Você precisa apenas alterar a ordem no `grid-template-areas`, alterando assim apenas o CSS sem alterar o HTML, desse modo:

Reviewer: ```suggestion
Tudo pronto! Essa definição ficou igual ao exemplo anterior onde não tinhamos o `grid-template-areas`, mas imagine que queiramos trocar a posição da `caixa-pequena2` pela `caixa-pequena4`. Você precisa apenas alterar a ordem no `grid-template-areas`, modificando assim apenas o CSS sem mexer no HTML, desse modo:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: 2fr repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-template-areas: 
    "c-pequena1 c-pequena4 c-pequena3"
    "c-pequena2 c-pequena5 c-pequena6";
}
```

# Hora de experimentar

> Agora que você já sabe um pouco de CSS Grid, brinque um pouco com essa propriedade e teste coisas que eu não expliquei nesse mini-guia. Ta sem ideia? Então bora lá: Se eu tiver uma área não utilizada no grid, o que acontece? E se eu quiser colocar uma área vazia entre cada coluna e cada linha? E se eu não souber a quantidade de elementos filhos, como eu faço?Tudo isso é possivel com o Grid, garanto a você.

Reviewer: ```suggestion
Agora que você já sabe um pouco de CSS Grid, brinque um pouco com essa propriedade e teste coisas que eu não expliquei nesse miniguia. Tá sem ideia? Então bora lá: Se eu tiver uma área não utilizada no grid, o que acontece? E se eu quiser colocar uma área vazia entre cada coluna e cada linha? E se eu não souber a quantidade de elementos filhos, como eu faço? Tudo isso é possivel com o Grid, garanto a você.
```

# É isto!

> Meu nome é Pedro e muito obrigado pela leitura! Sou contribuidor do OpenDevUFCG e desenvolvedor FrontEnd. Se desejar entrar em contato comigo só me mandar uma mensagem no [Twitter](https://twitter.com/pedro_espindula) e se quiser ver um pouco do meu trabalho, dá uma olhada no meu [GitHub](https://github.com/pedro_espindula)

Reviewer: ```suggestion
Meu nome é Pedro e muito obrigado pela leitura! Sou contribuidor do OpenDevUFCG e desenvolvedor front-end. Se desejar entrar em contato comigo, é só me mandar uma mensagem no [Twitter](https://twitter.com/pedro_espindula) e se quiser ver um pouco do meu trabalho, dá uma olhada no meu [GitHub](https://github.com/pedro_espindula).
```

---
title: Criando malhas simples com CSS Grid
published: false
description: Introdução a CSS grid
tags: css, grid, ptbr, tips
---

# Grid para ser feliz

Você conhece o CSS Grid? Se não, você está perdendo uma ótima ferramenta do CSS que facilita tanto o desenvolvimento como também a manutenção de código. Nesse post, irei tratar um pouco de seu funcionamento e de como ele pode ser poderoso.

# Um pouco sobre caixas

Pra começar, você tem que abstrair que todo elemento HTML está envolvido por uma caixa que tanto o delimita como também o seu conteúdo (para saber mais, acesse [esse link](https://tableless.github.io/iniciantes/manual/css/box-model.html) e [esse link](https://developer.mozilla.org/pt-BR/docs/Web/CSS/box_model)). Na maioria das vezes, vão existir elementos que têm dentro do seu conteúdo outros elementos HTML, ou seja, caixas dentro de caixas, e o que o Grid faz é simplesmente organizar essas caixas de acordo com sua vontade. Como ele faz isso? Ele define uma malha quadriculada e atribui espaços dessa malha a cada elemento HTML.

![](https://thepracticaldev.s3.amazonaws.com/i/58gv9hkeys3de94e0amo.png)

> *Fonte: [SparkBox](https://seesparkbox.com/foundry/css_grid_layout_guide_with_flexbox_fallbacks)*

Reviewer: ```suggestion
<figcaption>Fonte: <a href="https://seesparkbox.com/foundry/css_grid_layout_guide_with_flexbox_fallbacks">SparkBox</a></figcaption>
```
Sim, é estranho, mas o dev.to suporta HTML também (geralmente usa só pra legenda de imagem mesmo). Com isso vai ficar bonitinho lá embaixo da imagem, qualquer coisa vê a postagem de Lucas pra visualizar como vai ficar.

> A malha é definida por trackers que são basicamente delimitadores das linhas e colunas, que são esses marcadores brancos:

Reviewer: ```suggestion
A malha é definida por trackers, que são basicamente delimitadores das linhas e colunas, representados por esses marcadores brancos:
```

![](https://thepracticaldev.s3.amazonaws.com/i/kldxli86fo9f6mcgoi4r.jpg)

> *Fonte: [SparkBox](https://seesparkbox.com/foundry/css_grid_layout_guide_with_flexbox_fallbacks)*

Reviewer: ```suggestion
<figcaption>Fonte: <a href="https://seesparkbox.com/foundry/css_grid_layout_guide_with_flexbox_fallbacks">SparkBox</a></figcaption>
```

A linha é a área entre um tracker e outro de forma vertical, ou seja, cada novo espaço na altura da malha. Já a coluna é a área entre um tracker e outro de forma horizontal, ou seja, cada novo espaço no comprimento da malha.

As linhas e colunas são geralmente medidas em `fr`. Essa medida foi criada exclusivamente para o CSS Grid e pode ser definida como "uma fração do espaço disponível no container do grid" (fonte: [MDN](https://developer.mozilla.org/pt-BR/docs/Web/CSS/CSS_Grid_Layout/Basic_Concepts_of_Grid_Layout)). No caso da imagem acima, as colunas têm a medida de `5fr` e as linhas de `3fr`. E sabe o melhor do `fr`? Ele se adequa ao tamanho da sua tela. 

Além disso, cada caixa pintada da imagem é uma área do grid onde ficará um elemento HTML. Mas calma que eu vou explicar o porquê isso é importante.

# Colocando em prática

> Só pra você saber, os exemplos nesse post vão seguir uma estrutura onde mostrarei sempre um código css e uma aplicação desse código num codepen. Além disso, existirá um fluxo onde cada código será usado como base para os códigos subsequentes.

Reviewer: ```suggestion
Só pra você saber, os exemplos nesse post vão seguir uma estrutura onde mostrarei sempre um código CSS e uma aplicação desse código num codepen. Além disso, existirá um fluxo no qual cada código será usado como base para os códigos subsequentes.
```

## Preparando o terreno

O primeiro passo para utilizar o grid é definir quais são os elementos que estamos utilizando e o que cada um contém. Isso é definido no HTML. Nesse caso, temos uma caixa grande que tem seis caixas pequenas dentro dela:

```html
<div class="caixa grande">
  <div class="caixa pequena-1">Caixa pequena 1</div>
  <div class="caixa pequena-2">Caixa pequena 2</div>
  <div class="caixa pequena-3">Caixa pequena 3</div>
  <div class="caixa pequena-4">Caixa pequena 4</div>
  <div class="caixa pequena-5">Caixa pequena 5</div>
  <div class="caixa pequena-6">Caixa pequena 6</div>
</div>
```

Para facilitar o entendimento, pode-se usar tags HTML personalizadas. Segue o exemplo de como o código acima poderia ficar:

```html
<caixa-grande>
  <caixa-pequena1/>
  <caixa-pequena2/>
  <caixa-pequena3/>
  <caixa-pequena4/>
  <caixa-pequena5/>
  <caixa-pequena6/>
</caixa-grande>
```

> Para começar a utilização do grid, você precisa apenas definir a propriedade `display: grid` no elemento que desejar, no caso iremos utilizar essa propriedade na `caixa-grande` já que queremos organizar o que tem dentro dela: 

Reviewer: ```suggestion
Para começar a utilização do grid, você precisa apenas definir a propriedade `display: grid` no elemento que desejar. No caso iremos utilizar essa propriedade na `caixa-grande`, já que queremos organizar o que tem dentro dela: 
```

```css
caixa-grande {
  display: grid;
}

```

{% codepen https://codepen.io/pedroespindula/pen/mNQYGQ %}

Facil, né? Mas calma, ainda não temos nada pronto. O que foi definido nesse CSS foi apenas o tipo de organização (`display`) que a caixa utilizará, que é o grid. Nesse caso, estamos dizendo ao elemento apenas isso: "Use o grid pra organizar o que tem dentro de você. Não me importa como você vai fazer isso, apenas use o grid.".

# Entendendo como organizar

## Organização básica

> Você deve ter notado que a malha foi definida de forma automática onde cada caixa menor foi colocada em uma linha da malha, ou seja, um abaixo do outro. Implicitamente, ao definir o `display: grid` você define algumas outras propriedades padrões do grid ao elemento. A propriedade responsável por organizar o grid desse modo é a [grid-auto-flow](https://developer.mozilla.org/pt-BR/docs/Web/CSS/grid-auto-flow) que é definida de forma padrão como `grid-auto-flow: row`. Ou seja, o fluxo automático de organização do grid está definido como linha. Assim, a cada novo filho adicionado, ele vai ser posto numa nova linha.

Reviewer: ```suggestion
Você deve ter notado que a malha foi definida de forma automática onde cada caixa menor foi colocada em uma linha da malha, ou seja, um abaixo do outro. Implicitamente, ao definir o `display: grid`, você define algumas outras propriedades padrões do grid ao elemento. A propriedade responsável por organizar o grid desse modo é a [grid-auto-flow](https://developer.mozilla.org/pt-BR/docs/Web/CSS/grid-auto-flow), que é definida de forma padrão como `grid-auto-flow: row`. Ou seja, o fluxo automático de organização do grid está definido como linha. Assim, a cada novo filho adicionado, ele vai ser posto numa nova linha.
```

Agora imagine que queiramos organizar os elementos em colunas ao invés de linhas, ou seja, um ao lado do outro ao invés de um abaixo do outro. Basta modificar essa propriedade da seguinte forma:

```css
caixa-grande {
  display: grid;
  grid-auto-flow: column;
}
```
{% codepen https://codepen.io/pedroespindula/pen/qeLPzx %}

> Nesse exemplo estamos dizendo literalmente ao elemento "Use o grid mas organize essas caixas em colunas".

Reviewer: ```suggestion
Nesse exemplo estamos dizendo literalmente ao elemento "Use o grid, mas organize essas caixas em colunas".
```

## Organização com grid-template

> No exemplo anterior deixamos o trabalho da definição da malha todo para o grid de forma automatica. Mas e se quisermos definir uma malha personalizada? Para isso, precisamos definir o `grid-template` no elemento HTML de sua escolha. Nessa malha você vai definir três coisas: as colunas (columns), as linhas (rows) e as áreas (areas) para colocar o conteúdo da caixa que você escolher.

Reviewer: ```suggestion
No exemplo anterior deixamos o trabalho da definição da malha todo para o grid de forma automática. Mas e se quisermos definir uma malha personalizada? Para isso, precisamos definir o `grid-template` no elemento HTML de sua escolha. Nessa malha você vai definir três coisas: as colunas (columns), as linhas (rows) e as áreas (areas) para colocar o conteúdo da caixa que você escolher.
```

> Lembra da definição de linhas, colunas e áreas que eu dei la em cima? Pronto, você vai fazer seguir esses conceitos para definir sua malha.

Reviewer: ```suggestion
Lembra da definição de linhas, colunas e áreas que eu dei lá em cima? Pronto, você vai fazer seguir esses conceitos para definir sua malha.
```

## Codificando

> Vamos para um exemplo concreto. Imagine que você não queira utilizar o `grid-auto-flow` mas queira todas as caixas uma do lado da outra numa unica linha. Já que temos seis caixas pequenas, temos que definir seis colunas e uma linha, desse modo:

Reviewer: ```suggestion
Vamos para um exemplo concreto. Imagine que você não queira utilizar o `grid-auto-flow`, mas queira todas as caixas uma do lado da outra numa unica linha. Já que temos seis caixas pequenas, temos que definir seis colunas e uma linha, desse modo:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr;
}
```

{% codepen https://codepen.io/pedroespindula/pen/MNLZEV %}

> Já que temos apenas uma linha, não precisamos defini-la explicitamente por isso, para facilitar, a definição pode ser feita desse modo:

Reviewer: ```suggestion
Já que temos apenas uma linha, não precisamos defini-la explicitamente. Por isso, para facilitar, a definição pode ser feita desse modo:
```

```css
caixa-grande {
  display: grid;
  grid-template-columns: repeat(6, 1fr);
}
```

{% codepen https://codepen.io/pedroespindula/pen/VogNKJ %}

> Além da omissão do `grid-template-rows` utilizamos a função `repeat`. O `repeat` é utilizado para repetir várias vezes a quantidade de medida passada. Nesse caso, repetimos seis vezes a medida de `1fr`, já que todas as colunas teriam o mesmo tamanho de 1fr. O repeat é uma função muito poderosa e será abordada de forma mais aprofundada em outro post.

Reviewer: ```suggestion
Além da omissão do `grid-template-rows`, utilizamos a função `repeat`. O `repeat` é utilizado para repetir várias vezes a quantidade de medida passada. Nesse caso, repetimos seis vezes a medida de `1fr`, já que todas as colunas teriam o mesmo tamanho de 1fr. O repeat é uma função muito poderosa e será abordada de forma mais aprofundada em outro post.
```

> Agora se quisermos uma malha 3x2? Continua muito simples:

Reviewer: ```suggestion
Agora, e se quisermos uma malha 3x2? Continua muito simples:
```

Nele você define, literalmente, áreas que a malha terá. Como exemplo:

```css
caixa-grande {
  display: grid;
  grid-template-columns: 2fr repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-template-areas: 
    "c-pequena1 c-pequena2 c-pequena3"
    "c-pequena4 c-pequena5 c-pequena6";
}
```

O container pai está pronto, mas precisamos ainda definir dentro dos filhos a qual área cada um está atribuído, o que é feito desse modo:

```css
caixa-pequena1 {
  grid-area: c-pequena1;
}

caixa-pequena2 {
  grid-area: c-pequena2;
}

caixa-pequena3 {
  grid-area: c-pequena3;
}

caixa-pequena4 {
  grid-area: c-pequena4;
}

caixa-pequena5 {
  grid-area: c-pequena5;
}

caixa-pequena6 {
  grid-area: c-pequena6;
}
```

{% codepen https://codepen.io/pedroespindula/pen/VwZePoG %}

Tudo pronto! Essa definição ficou igual ao exemplo anterior onde não tinhamos o `grid-template-areas`, mas imagine que queiramos trocar a posição da `caixa-pequena2` pela `caixa-pequena4`. Você precisa apenas alterar a ordem no `grid-template-areas`, modificando assim apenas o CSS sem mexer no HTML, desse modo:

> Só uma observação, a definição da propriedade do `grid-area` é sem aspas, se você colocar a área envolta de aspas ele não vai funcionar.

Reviewer: Move essa observação pra depois do código, ficou meio estranho aqui

```css
caixa-grande {
  display: grid;
  grid-template-columns: 2fr repeat(2, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-template-areas: 
    "c-pequena1 c-pequena4 c-pequena3"
    "c-pequena2 c-pequena5 c-pequena6";
}
```

{% codepen https://codepen.io/pedroespindula/pen/YzKwZzR %}

> As áreas do grid não possibilitam somente a mudança de forma facil de elementos HTML de posição. A forma mais poderosa que ele pode ser usada é para a definição de malhas complexas. Uma coisa que não tinhamos feito antes é usar mais de uma vez uma área na definição do `grid-template-areas`, mas esse é o motivo principal de usarmos o grid.

Reviewer: ```suggestion
As áreas do grid não possibilitam somente a mudança fácil da posição de elementos HTML. A forma mais poderosa que ele pode ser usada é para a definição de malhas complexas. Uma coisa que não tínhamos feito antes é usar mais de uma vez uma área na definição do `grid-template-areas`, mas esse é o motivo principal de usarmos o grid.
```

> Para você entender melhor, imagine que temos agora que fazer uma pagina principal que tenha uma barra lateral, um conteúdo, um cabeçalho e um rodapé. O cabeçalho deve ficar sempre em cima, a barra lateral no lado esquerdo, o conteúdo principal no lado direito e o rodapé sempre em baixo da barra lateral. O conteúdo principal é três vezes maior que a barra lateral em largura e cinco vezes maior que o cabeçalho e o rodapé em altura. 

Reviewer: ```suggestion
Para você entender melhor, imagine que temos agora que fazer uma página principal que tenha uma barra lateral, um conteúdo, um cabeçalho e um rodapé. O cabeçalho deve ficar sempre em cima, a barra lateral no lado esquerdo, o conteúdo principal no lado direito e o rodapé sempre em baixo da barra lateral. O conteúdo principal é três vezes maior que a barra lateral em largura e cinco vezes maior que o cabeçalho e o rodapé em altura. 
```

> Parece complexo né? É não, isso fica fácil de organizar com grid e com grid-areas. Aconselho você a tentar fazer esse exemplo primeiro antes de ver o jeito que eu fiz. Já fez? Ótimo! Não conseguiu? Sem problemas, a gente tá aqui pra isso! Nem tentou? Tou de olho hein!? 

Reviewer: ```suggestion
Parece complexo, né? Não é; isso fica fácil de organizar com grid e com grid-areas. Aconselho você a tentar fazer esse exemplo primeiro antes de ver o jeito que eu fiz. Já fez? Ótimo! Não conseguiu? Sem problemas, a gente tá aqui pra isso! Nem tentou? Tô de olho, hein!? 
```

Reviewer: KKKKKKKKKKKKKKKKKKKKKKKKKKKKKK julgando o leitor

Author: :eyes:

> Vamos a resolução primeiro definimos todos os elementos:

Reviewer: ```suggestion
Vamos à resolução: primeiro, definimos todos os elementos:
```

```html
<pagina-principal>
  <cabecalho/>
  <barra-lateral/>
  <conteudo-principal/>
  <rodape/>
</pagina-principal>
```

Depois o grid:

```css
pagina-principal {
  display: grid;
}
```

> Eu normalmente defino o `grid-template-areas` primeiro pois assim tenho uma visualização melhor do que estou tentando fazer:

Reviewer: ```suggestion
Eu normalmente defino o `grid-template-areas` antes, pois assim tenho uma visualização melhor do que estou tentando fazer:
```

```css
pagina-principal {
  display: grid;
  grid-template-areas: 
    "cabecalho cabecalho"
    "barra-lateral conteudo-principal"
    "rodape conteudo-principal";
}
```

> Depois podemos atribuir cada área a um elemento filho:

Reviewer: ```suggestion
Depois, podemos atribuir cada área a um elemento filho:
```

```css
cabecalho {
  grid-area: cabecalho;
}

barra-lateral {
  grid-area: barra-lateral;
}

conteudo-principal {
  grid-area: conteudo-principal;
}

rodape {
  grid-area: rodape;
}

```

> Mas ainda falta definir os tamanhos que foram pedidos, pra isso usamos a definição das linhas e das colunas. Como o conteúdo principal vai ser três vezes maior que a barra lateral, teremos uma coluna três vezes maior que a outra, e como o conteúdo principal vai ser quatro vezes maior que o cabeçalho e o rodapé, a linha do conteúdo principal vai ter que ser quatro vezes maior que as linhas do cabeçalho e do rodapé:

Reviewer: ```suggestion
Mas ainda falta definir os tamanhos que foram pedidos; pra isso, usamos a definição das linhas e das colunas. Como o conteúdo principal vai ser três vezes maior que a barra lateral, teremos uma coluna três vezes maior que a outra, e como o conteúdo principal vai ser quatro vezes maior que o cabeçalho e o rodapé, a linha do conteúdo principal vai ter que ser quatro vezes maior que as linhas do cabeçalho e do rodapé:
```

```css
pagina-principal {
  display: grid;
  grid-template-columns: 1fr 3fr;
  grid-template-rows: 1fr 4fr 1fr;
  grid-template-areas: 
    "cabecalho cabecalho"
    "barra-lateral conteudo-principal"
    "rodape conteudo-principal";
}
```

{% codepen https://codepen.io/pedroespindula/pen/gOYPmrN %}

> Pronto, temos a nossa pagina principal feita!

Reviewer: ```suggestion
Pronto, temos a nossa página principal feita!
```
